src/coreclr/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -2133,47 +2133,74 @@ void CodeGen::genFMAIntrinsic(GenTreeHWIntrinsic* node)
     // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
     assert(!copiesUpperBits || !op1->isContained());
 
-    if (op2->isContained() || op2->isUsedFromSpillTemp())
+    unsigned overwrittenOpNum = 0;
+    LIR::Use use;
+    if (LIR::AsRange(compiler->compCurBB).TryGetUse(node, &use))
     {
-        // 132 form: op1 = (op1 * op3) + [op2]
+        overwrittenOpNum = node->GetOverwrittenOpNumForFMA(use.User(), op1, op2, op3);
+    }
 
-        ins    = (instruction)(ins - 1);
-        op1Reg = op1->GetRegNum();
-        op2Reg = op3->GetRegNum();
-        op3    = op2;
+    if (overwrittenOpNum == 1)
+    {
+        if (op2->isContained())
+        {
+            // op1 = (op1 * [op2]) + op3
+            // 132 form: XMM1 = (XMM1 * [XMM3]) + XMM2
+            ins    = (instruction)(ins - 1);
+            op1Reg = op1->GetRegNum();
+            op2Reg = op3->GetRegNum();
+            op3    = op2;
+        }
+        else
+        {
+            // op1 = (op1 * op2) + [op3]
+            // 213 form: XMM1 = (XMM2 * XMM1) + [XMM3]
+            op1Reg = op1->GetRegNum();
+            op2Reg = op2->GetRegNum();
+        }
     }
-    else if (op1->isContained() || op1->isUsedFromSpillTemp())
+    else if (overwrittenOpNum == 3)
     {
-        // 231 form: op3 = (op2 * op3) + [op1]
-
+        // 231 form: XMM1 = (XMM2 * [XMM3]) + XMM1
+        // One of the following:
+        // op3 = ([op1] * op2) + op3
+        // op3 = (op1 * [op2]) + op3
         ins    = (instruction)(ins + 1);
         op1Reg = op3->GetRegNum();
-        op2Reg = op2->GetRegNum();
-        op3    = op1;
+        if (op1->isContained())
+        {
+            // op3 = ([op1] * op2) + op3
+            op2Reg = op2->GetRegNum();
+            op3    = op1;
+        }
+        else
+        {
+            // op3 = (op1 * [op2]) + op3
+            op2Reg = op1->GetRegNum();
+            op3    = op2;
+        }
     }
     else
     {
-        // 213 form: op1 = (op2 * op1) + [op3]
-
-        op1Reg = op1->GetRegNum();
-        op2Reg = op2->GetRegNum();
-
-        isCommutative = !copiesUpperBits;
-    }
-
-    if (isCommutative && (op1Reg != targetReg) && (op2Reg == targetReg))
-    {
-        assert(node->isRMWHWIntrinsic(compiler));
+        assert(overwrittenOpNum == 2 || overwrittenOpNum == 0);
```

Comment:
I think `overittenOpNum == 0` should/could be its own path.

This doesn't, for example, currently handle the case where `(overwrittenOpNum == 0) && op2->isContained()` in which case we want to swap op1/op2 and still use the `132 form`.